scipy/stats/_survival.py

```diff
@@ -374,3 +384,122 @@ def _ecdf_right_censored(sample):
+
+
+def log_rank(sample: CensoredData, control: CensoredData) -> LogRankResult:
+    """Log Rank test.
+
+    Parameters
+    ----------
+    sample, control : CensoredData
+        Sample and control data to compare based on their survival functions.
+        The order does not matter, `sample` and `control` can be interverted.
+
+    Returns
+    -------
+    res : LogRankResult
+        An object containing attributes:
+
+        statistic : float ndarray
+            The computed statistic of the test.
+        pvalue : float ndarray
+            The computed p-value of the test.
+
+    References
+    ----------
+    .. [1] ...
+
+    Examples
+    --------
+
+    >>> from scipy import stats
+    >>> sample = stats.CensoredData(
+    ...     uncensored=[ 8., 12., 26., 14., 21., 27.],
+    ...     right=[ 8., 32., 20., 40.]
+    ... )
+    >>> control = stats.CensoredData(
+    ...     uncensored=[33., 28., 41.],
+    ...     right=[48., 48., 25., 37., 48., 25., 43.]
+    ... )
+
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> ax = plt.subplot()
+    >>> ecdf_sample = stats.ecdf(sample)
+    >>> ax.step(np.insert(ecdf_sample.x, 0, 0), np.insert(ecdf_sample.sf.points, 0, 1),
+    ...         where='post', label='sample')
+    >>> ecdf_control = stats.ecdf(control)
+    >>> ax.step(np.insert(ecdf_control.x, 0, 0), np.insert(ecdf_control.sf.points, 0, 1),
+    ...         ls='-.', where='post', label='control')
+    >>> ax.set_xlabel('Time (months)')
+    >>> ax.set_ylabel('Empirical CDF')
+    >>> plt.legend()
+    >>> plt.show()
+
+    >>> res = stats.log_rank(sample=sample, control=control)
+    >>> res.statistic
+    6.148087536256203
+    >>> res.pvalue
+    0.013155428547469983
+
+    """
+    sample, control = _iv_log_rank(sample=sample, control=control)
+
+    sample_tot = CensoredData(
+        uncensored=np.concatenate((sample._uncensored, control._uncensored)),
+        right=np.concatenate((sample._right, control._right))
+    )
+
+    res = ecdf(sample_tot)
+    times_sample_tot = res.x[res.sf._d.astype(bool)]
+    at_risk_sample_tot = res.sf._n[res.sf._d.astype(bool)]
+    _, dead_per_times = np.unique(sample_tot._uncensored, return_counts=True)
+
+    n_died_sample = sample._uncensored.size
+    n_died_control = control._uncensored.size
+
+    times = np.sort(np.concatenate((sample._uncensored, sample._right)))
+    at_risk_sample = _at_risk(times, times_sample_tot)
+
+    times = np.sort(np.concatenate((control._uncensored, control._right)))
+    at_risk_control = _at_risk(times, times_sample_tot)
+
+    sum_exp_event_sample = np.sum(at_risk_sample * (dead_per_times/at_risk_sample_tot))
+    sum_exp_event_control = np.sum(at_risk_control * (dead_per_times/at_risk_sample_tot))
+
+    statistic = (
+        (n_died_sample - sum_exp_event_sample)**2/sum_exp_event_sample
+        + (n_died_control - sum_exp_event_control)**2/sum_exp_event_control
+    )
+
+    pvalue = 1 - chi2(df=1).cdf(statistic)
```

Comment:
```suggestion
    pvalue = chi2(df=1).sf(statistic)
```
But there are other versions of the test that use a slightly different statistic and would let you bring back `alternative`, such as the one on Wikipedia.

![image](https://user-images.githubusercontent.com/6570539/227105036-ce40a9b8-00ba-42e8-94f6-482f8fdd8f55.png)

Reply:
Cool, but then the question is which implementation do we want to have? The standard version, or something else?

Reply:
Depends on what is considered to be standard. But actually, when I did this sort of thing before, the p-values turned out to be identical for the two-sided case. It just allows you to define one-sided alternatives. We can leave it alone for now but consider a follow-up. (Pretty much all other hypothesis tests have one-sided alternatives if there is a reasonable way to define it.)

Reply:
Yeah this suggests that we'll get the same p-value for two-sample.

![image](https://user-images.githubusercontent.com/6570539/227372665-b4dac23b-b893-4157-8588-3507a53a7cdc.png)